src/wasm/wasm-type.cpp

```diff
@@ -104,38 +172,79 @@ void Type::init(const std::vector<Type>& types) {
     *this = types[0];
     return;
   }
+  id = canonicalize(TypeDef(tuple));
+}
 
-  // Add a new type if it hasn't been added concurrently
-  std::lock_guard<std::mutex> lock(mutex);
-  auto indexIt = indices.find(types);
-  if (indexIt != indices.end()) {
-    id = indexIt->second;
+Type::Type(const Signature signature, bool nullable) {
+  id = canonicalize(TypeDef(signature, nullable));
+}
+
+Type::Type(const Struct& struct_, bool nullable) {
+#ifndef NDEBUG
+  for (Field f : struct_.fields) {
+    assert(f.type.isSingle() && f.type.isConcrete());
+  }
+#endif
+  id = canonicalize(TypeDef(struct_, nullable));
+}
+
+Type::Type(const Array& array, bool nullable) {
+#ifndef NDEBUG
+  assert(array.element.type.isSingle() && array.element.type.isConcrete());
+#endif
+  id = canonicalize(TypeDef(array, nullable));
+}
+
+bool Type::isTuple() const {
+  if (id <= _last_basic_id) {
+    return false;
   } else {
-    auto vec = std::make_unique<std::vector<Type>>(types);
-    id = uintptr_t(vec.get());
-    constructedTypes.push_back(std::move(vec));
-    assert(id > _last_value_type);
-    indices[types] = id;
+    auto* typeDef = (TypeDef*)id;
+    return typeDef->isTuple();
   }
 }
 
-Type::Type(std::initializer_list<Type> types) { init(types); }
+bool Type::isRef() const {
+  if (id <= _last_basic_id) {
+    return id >= funcref && id <= exnref;
+  } else {
+    auto* typeDef = (TypeDef*)id;
+    switch (typeDef->kind) {
+      case TypeDef::TupleKind:
+        return false;
+      case TypeDef::SignatureKind:
+      case TypeDef::StructKind:
+      case TypeDef::ArrayKind:
+        return true;
+    }
+    WASM_UNREACHABLE("unexpected kind");
+  }
+}
 
-Type::Type(const std::vector<Type>& types) { init(types); }
+bool Type::isNullable() const {
+  if (id <= _last_basic_id) {
+    return id >= funcref && id <= exnref;
+  } else {
+    auto* typeDef = (TypeDef*)id;
+    return typeDef->isNullable();
+  }
+}
 
 size_t Type::size() const { return expand().size(); }
 
-const std::vector<Type>& Type::expand() const {
-  if (id <= _last_value_type) {
-    return basicTypes[id];
+const TypeList& Type::expand() const {
+  if (id <= Type::_last_basic_id) {
+    return basicTuples[id];
   } else {
-    return *(std::vector<Type>*)id;
+    auto* typeDef = (TypeDef*)id;
+    assert(typeDef->isTuple() && "can only expand tuple types");
```

Comment:
I anticipate this restriction causing a bunch of problems. As a follow-up PR we should probably switch from using `expand` to an iterator-based approach so that all types can be treated uniformly.

Reply:
Yep, just ran into this when trying to create a function with a single array parameter:

```cpp
Signature(Type(array, false), Type::none)
```

Goes boom when trying to print the signature.

Reply:
If this basically kills all the interesting tests, it might make sense to split this change off into a preliminary PR instead.

Reply:
And there we go again :) Can you paste me a couple lines of code how an iterator-based approach might look like, just so I have a starting point?

Reply:
Actually this is one of the most terrible things in C++, IMO :(

Basically you need to create a new struct that supports all of the necessary iterator operations so it can be used anywhere an iterator is expected. There are many different types of iterators that each support slightly different sets of operations, but we should use a _LegacyRandomAccessIterator_, which what std::vectors use. The docs are here: https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator. You can see another example of an iterator implemented in Binaryen here: https://github.com/WebAssembly/binaryen/blob/5d7248b1613017af4f26d176e6ee725210b0fa88/src/support/small_vector.h#L132-L174, although that one is a little complicated because it does shenanigans to share code between the const iterator and the normal iterator.

For this use case, we only need a constant iterator because Types are immutable. The iterator should know whether the type is iterating over is a tuple, a single type, or an empty type. If the type is a tuple, the iterator should just forward all of its operations to a const iterator of the `TypeList` backing the tuple. Otherwise, the iterator should implement the operations itself. For single types, `begin` should return an iterator that just dereferences to that type and `end` should be some sentinel value. For empty types, `begin` and `end` should both be that sentinel value.

Let me know if you want more details or if you would prefer that I tackle this one :)

Reply:
Thanks, sounds interesting, going to look into it :)

Reply:
Got it working, now replacing all `expand()`s with iterator usage. Couple questions:

* What's the best way to convert a class with a custom iterator implementation to a `std::vector<TheClass>`? That's required in some parts of the code, but one can't just do `std::vector<Type>(type.begin(), type.end())` because the iterator is custom, so I'm `push_back`ing in a loop, which feels meh.
* Is there a better way than incrementing a separate `size_t index` when iterating `for (auto t : type) { ... ++index; }` where inner code needs an index?

Reply:
I think carrying an explicit index around is fine. The alternative would be to implement the iterator as yet another tagged union, where the cases are:

 1. The type is a tuple, so the iterator just contains a `std::vector<Type>::const_iterator` and forwards all methods to it.

 2. The type is a singleton, so the iterator just contains the type and can return a const reference to it. Incrementing this iterator puts it into the next state:

 3. Sentinel value used for iterators into empty types (i.e. none or unreachable) and used as the `end()` for singleton types.

But clearly just keeping an index like you did is much simpler 👍

Reply:
Yay https://github.com/WebAssembly/binaryen/pull/2510#discussion_r355432470